### ISSUE TITLE
Different Vector Length on X86_64 causes segmentation fault in TypeDictionary of a consuming project.

### ISSUE THREAD

**Issue by georgkrylov** [eclipse/omr#6544]
Dear OMR users and developers, I have ran into an issue with one of the recent commits:
I found that my code segfaults when I have the following commit : https://github.com/eclipse/omr/commit/53c6333ab27beeccba393dd1d9ea79997aebf927
I am not certain what exactly is the cause of the issue, but the commit breaks my runtime that was otherwise working. The segFault is in inherited section of code, when creating  a TypeDictionary.

The approximate stack trace using public repos:
https://github.com/eclipse/omr/blob/53c6333ab27beeccba393dd1d9ea79997aebf927/compiler/ilgen/OMRTypeDictionary.cpp#L344
https://github.com/eclipse/omr/blob/53c6333ab27beeccba393dd1d9ea79997aebf927/compiler/ilgen/OMRTypeDictionary.cpp#L598
https://github.com/wasmjit-omr/wasmjit-omr/blob/072f003f1425035b62fd1bd77457f8511031078c/src/jit/type-dictionary.cc#L32
(This is not my repo, but I forked this repo and never touched TypeDictionary stuff yet)

For now, I have a workaround to revert this commit when doing my rebases, as I do have changes that will never be added to OMR but I need them for my runtime to work. However, in the long run, I am worried if the development continues on that commit it will become a burden to maintain the workaround.

I "blame" this ``#ifdef``, as removing it makes my runtime "happy"
https://github.com/eclipse/omr/blob/53c6333ab27beeccba393dd1d9ea79997aebf927/compiler/il/OMRDataTypes.hpp#L228

Can anyone please suggest a fix on how to maybe update https://github.com/wasmjit-omr/wasmjit-omr/blob/072f003f1425035b62fd1bd77457f8511031078c/src/jit/type-dictionary.cc#L32 , to be up to date with the commit being present in OMR codebase. Or maybe identify what is the issue and how come vector length enum breaks the TypeDictionary (albeit hacky, if reading comments).

My apologies if this is an issue of a consuming project.

**Comment by gita-omr** [comment]
Thanks for describing the problem. I think, in this case, it's important to take a look at this PR: https://github.com/eclipse/omr/pull/6353 and make all the necessary adjustments in your project. Basically, vector types moved to another place and all tables indexed by `DataTypes` enum need to be adjusted. In addition, the overall number of types increases with `NumVectorLengths = VectorLength512`. If corresponding code is not adjusted it might cause some issues, like accessing bad memory.

Please let me know if you have any questions or no apparent issues have been found.

**Comment by mstoodle** [comment]
Actually, looking at the wasmjit-omr code you linked to: `toIlType<decltype(Value::i32)>(this))` would normally be written as: `this->toIlType<decltype(Value::i32)>()`. I'm not sure if that's the source of the problem, though.

**Comment by mstoodle** [comment]
The result you're seeing may just be `toIlType<>()` not knowing what `TR::IlType` corresponds to `wabt::jit::TypeDictionary *`

**Comment by mstoodle** [comment]
(deleted a foolish comment suggesting an explanation that did not make sense)

**Comment by mstoodle** [comment]
No idea why that #ifdef would affect how this code works, though. That's quite strange.

**Comment by gita-omr** [comment]
I have a feeling that adding more vector types (by having` NumVectorLengths = VectorLength512` instead of `NumVectorLengths = VectorLength128`) caused some inconsistencies in the tables indexed by the `DataTypes` enum. I think we would need to search wasmjit-omr source for any such table.

**Comment by georgkrylov** [comment]
Thank you @mstoodle and @gita-omr .

To add more information, as per requested, instructions to reproduce the segfault:
To build a working runtime 
```
git clone git@github.com:CAS-Atlantic/wabtaot.git --branch main --single-branch wabtaot
cd wabtaot
git submodule update --init --recursive .
mkdir build
sed -i '37 i #include <time.h>' third_party/omr/compiler/cs2/timer.h
cd build
cmake ..
make -j12
cd ..
./bin/wat2wasm test/jit/call.txt -o call.wasm
./bin/wasm-interp call.wasm --jit-threshold=0 --run-all-exports
```
To build it with the change that broke it:

```
git clone git@github.com:CAS-Atlantic/wabtaot.git --branch main --single-branch wabtaot
cd wabtaot
git submodule update --init --recursive .
mkdir build
cd third_party/omr
git pull origin master
cd ../..
sed -i '37 i #include <time.h>' third_party/omr/compiler/cs2/timer.h
cd build
cmake -DCMAKE_BUILD_TYPE=Debug ..
make -j12
cd ..
./bin/wat2wasm test/jit/call.txt -o call.wasm
gdb --args ./bin/wasm-interp call.wasm --jit-threshold=0 --run-all-exports
```
The comment about
``this->toIlType<decltype(Value::i32)>()``
did not build successfully

**Comment by gita-omr** [comment]
Thanks @georgkrylov ! I will try to reproduce soon.

**Comment by gita-omr** [comment]
After some debugging, I think I finally figured it out. The problem is that OMR files are compiled with `TR_TARGET_X8` defined. While wasm files are compiled without it. This causes `TR::NumVectorLengths `to be defined differently in different files and therefore creates wrong view of some data (TypeDictionary::Int32 for example).

**Comment by gita-omr** [comment]
We wanted `TR::NumVectorLengths` to be defined based on project and platform. Perhaps there is a better way to do it... 

@BradleyWood @0xdaryl : I wonder what you think.

**Comment by georgkrylov** [comment]
Thank you @gita-omr . Indeed, adding `set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -DTR_TARGET_X86  ")` to my root `CMakeLists.txt` resolved the segmentation fault. I am now curious if this action has now became a hidden undocumented requirement on any project that consumes jitbuilder, or if I am consuming jitbuilder wrongly.

**Comment by gita-omr** [comment]
> Thank you @gita-omr . Indeed, adding `set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -DTR_TARGET_X86 ")` to my root `CMakeLists.txt` resolved the segmentation fault. I am now curious if this action has now became a hidden undocumented requirement on any project that consumes jitbuilder, or if I am consuming jitbuilder wrongly.

This is a very good question. Perhaps a better solution would be if we could  include a new header file and each platform/project could override it with its own. That header file would have only the definition of `NumVectorLengths` specific to that platform/project.

**Comment by mstoodle** [comment]
Something isn't right. Variables like TR_TARGET_X86 or LINUX have to be set to build the compiler component properly; if they aren't being set, then wasmjit-omr isn't building OMR properly and lots of things will break in unexpected ways. I strongly suspect that if you build OMR independently and simply build wasmjit-omr with the libjitbuilder.a static library that's created, then wasmjit-omr should work just fine.  Unfortunately, I'm not enough of a CMake expert to know why what you've done in wasmjit-omr isn't including and/or configuring OMR properly. I could very easily be wrong, but this line in your `CMakeLists.txt` (along with some of the other things to hack the cmake module path and to specifically include only `omrTargetSupport`) looks suspicious to me:
```
add_subdirectory(third_party/omr EXCLUDE_FROM_ALL)
```

I'm sure this is a naive question (see earlier "I'm not enough of a CMake expert" confession), but why do you need the other bits in addition to just `add_subdirectory(third_party/omr)`? Wouldn't that be the usual way to bring OMR's full build configuration into a consuming project?

The only CMake instructions for OMR I'm aware of are here: https://github.com/eclipse/omr/blob/master/doc/BuildingWithCMake.md . But this document does not describe how a consuming project could bring OMR's cmake files into its own build, as I believe wasmjit-omr is doing.

**Comment by georgkrylov** [comment]
I second  you on "The only CMake instructions for OMR I'm aware of are here: https://github.com/eclipse/omr/blob/master/doc/BuildingWithCMake.md . But this document does not describe how a consuming project could bring OMR's cmake files into its own build, as I believe wasmjit-omr is doing. If there's an expected (hopefully noncomplicated) way to do that, we should update that file to describe how to do it." that, @mstoodle . This document is practically useless for people who want to do something beyond running `make test` outside omr/openj9 infrastructure. Maybe it's also worth to resurrect this issue, while we are at it: https://github.com/eclipse/omr/issues/5786

I also wish I could just write `add_subdirectory(third_party/omr EXCLUDE_FROM_ALL)` 
but one commit broke this functionality introducing `add_omr_executable`. https://github.com/eclipse/omr/commit/222c0d0719dc1eb10c400ad7c2e03777e52a48b8 .
Default CMAKE is unhappy about it, unless you introduce it by showing OmrTargetSupport to it. I am not a CMAKE expert either, maybe I am supposed to use CMAKE package OMR constructs. I tried tweaking my project to use the package, but it's beyond my level of  abilities and isn't research I am  doing per se. So far these two lines seem the shortest and the cleanest way to consume OMR within a project that is not OpenJ9. 
Specific issue that I encountered is that `add_subdirectory(third_party/omr EXCLUDE_FROM_ALL)`  requires the "glue" target to be built, that uses `omr_add_executable` that is defined  by  `add_subdirectory(third_party/omr EXCLUDE_FROM_ALL)`  that  requires the "glue" target ....

I have inherited `wasmjit-omr` as-is and I had to reconfigure most of the includes manually to work on my compiler. In my prototype implementation (that is not ready to become public) I had to duplicate some CMAKE code (like showing platforms, targets etc. to my compiler folder). As my project uses extensible classes, one solution I found is to add my files to OMR targets as well as to my project targets, allowing shadowing OMR methods with TR_UNIMPLEMENTED , as well as extending INCLUDE_DIRS to configure IPATH. Maybe there's a better way to use OMR, but I am not aware. This paragraph is just to demonstrate how 'simple' it is to consume OMR.

**Comment by mstoodle** [comment]
Have you tried looking at how OpenJ9 consumes OMR? It may be overkill for what you're trying to do, but maybe there's something to learn from another example?

**Comment by georgkrylov** [comment]
@mstoodle @gita-omr Thank you for your advise. I am currently exploring `create_omr_compiler_library`  in conjunction with `make_compiler_target`. These two commands seem to provide definitions and includes. I am curious if that was the intent for the developers who consume jitbuilder, to follow  the example of `testcompiler` target.

**Comment by georgkrylov** [comment]
I've learned what i needed to do to consume OMR jitbuilder properly, but for other users it might be unclear. 

Just maybe to warn everyone, there's another bug I've uncovered in the process of learning how to propagate all the defines added in TR:
If someone wants to use ```TR_ASSERT``` but only links with, say jitbuilder, as `wasmjit-omr` was doing , without providing the flags like `-DLINUX` and now `-DTR_TARGET_ARCH_X86` when building the consuming project and, specifically for this issue I am currently pointing at `DEBUG` or `PROD_WITH_ASSUMES`, none of the `TR_ASSERT`s like 
https://github.com/CAS-Atlantic/wabtaot/blob/c8ad9a07ae8d1ff4647a66fe1d761ee998b32de7/src/jit/stack.cc#L42
would actually fire.

That's because without these flags TR_ASSERT is a no-op?
https://github.com/eclipse/omr/blob/e287e61337dbf37ab2929a102b59ad0687260aa4/compiler/infra/Assert.hpp#L233
